File: c++/2.cpp
#include <fstream>
#include <sstream>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

bool areValid(const vector<int>& reports, const vector<int>& sortedReports) {
    int previous = reports[0];
    for (int i = 1; i < reports.size(); i++) {
        if (reports[i] != sortedReports[i]) return false;
        if (int difference = abs(reports[i] - previous); difference > 3 || difference < 1) return false;
        previous = reports[i];
    }

    return true;
}

int main() {
    int save = 0;
    ifstream input{"../../data/2.txt"};

    string line;
    while (getline(input, line)) {
        vector<int> reports;

        istringstream stream(line);
        string word;
        while (stream >> word) {
            reports.push_back(stoi(word));
        }

        vector<int> sortedReports(reports);
        sort(sortedReports.begin(), sortedReports.end());
        if (areValid(reports, sortedReports)) {
            save += 1;
            continue;
        }

        reverse(sortedReports.begin(), sortedReports.end());
        if (!areValid(reports, sortedReports)) continue;
        save += 1;
    }

    cout << to_string(save) << endl;
}

File: python/2.py
def safe(reports: list) -> bool:
    sorted_reports = sorted(reports)

    if not (reports == sorted_reports or reports == list(reversed(sorted_reports))):
        return False

    return not any(a == b or abs(a - b) > 3 for a, b in zip(reports, reports[1:]))

part_1, part_2 = 0, 0
for line in open("../data/2.txt").read().splitlines():
    reports = list(map(int, line.split(" ")))

    if safe(reports):
        part_1 += 1
        part_2 += 1
        continue

    for item in range(len(line)):
        if safe(reports[:item] + reports[item + 1:]):
            part_2 += 1
            break

print(part_1)
# >>> 369
print(part_2)
# >>> 428